src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs

```diff
@@ -0,0 +1,396 @@
+                    Object.ReferenceEquals(value1.Module, value2.Module);
+            }
+            protected override BoxedValueType CreateValueFromKey(BoxedValuetypeHashtableKey key)
+            {
+                return new BoxedValueType(key.OwningModule, key.ValueType);
+            }
+        }
+        private BoxedValuetypeHashtable _boxedValuetypeHashtable = new BoxedValuetypeHashtable();
+
+        private struct UnboxingThunkHashtableKey
+        {
+            public readonly MethodDesc TargetMethod;
+            public readonly BoxedValueType OwningType;
+
+            public UnboxingThunkHashtableKey(MethodDesc targetMethod, BoxedValueType owningType)
+            {
+                TargetMethod = targetMethod;
+                OwningType = owningType;
+            }
+        }
+
+        private class UnboxingThunkHashtable : LockFreeReaderHashtable<UnboxingThunkHashtableKey, GenericUnboxingThunk>
+        {
+            protected override int GetKeyHashCode(UnboxingThunkHashtableKey key)
+            {
+                return key.TargetMethod.GetHashCode();
+            }
+            protected override int GetValueHashCode(GenericUnboxingThunk value)
+            {
+                return value.TargetMethod.GetHashCode();
+            }
+            protected override bool CompareKeyToValue(UnboxingThunkHashtableKey key, GenericUnboxingThunk value)
+            {
+                return Object.ReferenceEquals(key.TargetMethod, value.TargetMethod) &&
+                    Object.ReferenceEquals(key.OwningType, value.OwningType);
+            }
+            protected override bool CompareValueToValue(GenericUnboxingThunk value1, GenericUnboxingThunk value2)
+            {
+                return Object.ReferenceEquals(value1.TargetMethod, value2.TargetMethod) &&
+                    Object.ReferenceEquals(value1.OwningType, value2.OwningType);
+            }
+            protected override GenericUnboxingThunk CreateValueFromKey(UnboxingThunkHashtableKey key)
+            {
+                return new GenericUnboxingThunk(key.OwningType, key.TargetMethod);
+            }
+        }
+        private UnboxingThunkHashtable _unboxingThunkHashtable = new UnboxingThunkHashtable();
+        
+
+        /// <summary>
+        /// A type with an identical layout to the layout of a boxed value type.
+        /// </summary>
+        private class BoxedValueType : MetadataType
+        {
+            private const string BoxedValueFieldName = "BoxedValue";
+
+            public FieldDesc BoxedValue { get; }
+
+            public MetadataType ValueTypeRepresented { get; }
+
+            public override ModuleDesc Module { get; }
+
+            public override string Name => "Boxed_" + ValueTypeRepresented.Name;
+
+            public override string Namespace
+            {
+                get
+                {
+                    StringBuilder sb = new StringBuilder();
```

Comment:
I'm wondering if I should just move this all to `NodeFactory` so that I have access to `NameMangler` and can generate stable non-conflicting names. The only reason why this cares about names is to generate a unique symbol.